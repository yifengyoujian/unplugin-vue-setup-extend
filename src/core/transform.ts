import { parse, compileScript } from '@vue/compiler-sfc'
import MagicString from 'magic-string'
import { basename } from "path"

// fix: Remove the vue-demi  
const generateScript = (lang: string | true, name: string | true): string =>
  `<script${lang ? ` lang="${lang}"` : ''
  }> export default {name: '${name}'}</script>`

export default function transform(code: string, id: string) {
  const { descriptor } = parse(code)

  // fix: No scriptSetup or have a script to exit directly
  if (!descriptor.scriptSetup || descriptor.script) {
    return null
  }

  const magic = new MagicString(code)

  let SFCScriptBlock;
  try {
    SFCScriptBlock = compileScript(descriptor, { id })
  } catch {
    return null
  }

  const {
    attrs: { name, lang },
  } = SFCScriptBlock

  name && magic.appendLeft(0, generateScript(lang, name))

  const map = magic.generateMap({ hires: true })
  const filename = basename(id)

  map.file = filename
  map.sources = [filename]

  return {
    map,
    code: magic.toString(),
  }
}