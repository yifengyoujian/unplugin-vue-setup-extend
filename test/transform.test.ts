import { describe, expect, it } from 'vitest'
import transform from '../src/core/transform'

describe('transform', () => {
  it('transform script', async () => {
    const code = `<script setup name="App">// placeholder </script>`

    expect(transform(code, 'App.vue')!.code)
  })

  it('Add lang attributes quickly to transformed code', async () => {
    const code = `<script setup lang="ts" name="App">// placeholder </script>`
 
    expect(transform(code, 'App.vue')!.code).toMatchSnapshot()
  })

  it('If there is a script that does not convert', async () => {
    const code =
      `<script lang="ts">// placeholder </script>
       <script lang="ts" setup name="App">// placeholder </script>`

    expect(transform(code, 'App.vue')).toEqual(null)
  })
})
