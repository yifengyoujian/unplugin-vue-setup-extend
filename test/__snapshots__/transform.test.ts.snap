// Vitest Snapshot v1

exports[`transform > transform script 1`] = `null`;

exports[`transform > transform script js 1`] = `
{
  "code": "<script> export default {name: 'App'}</script><script setup name=\\"App\\">// placeholder </script>",
  "map": SourceMap {
    "file": "",
    "mappings": "8CAAA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": [
      null,
    ],
    "version": 3,
  },
}
`;

exports[`transform > transform script setup 1`] = `
{
  "code": "<script lang=\\"ts\\"> export default {name: 'App'}</script><script setup lang=\\"ts\\" name=\\"App\\">// placeholder </script>",
  "map": SourceMap {
    "file": "",
    "mappings": "wDAAA,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC,CAAC",
    "names": [],
    "sources": [
      "",
    ],
    "sourcesContent": [
      null,
    ],
    "version": 3,
  },
}
`;
